power_grid_model_c/power_grid_model/include/power_grid_model/grouped_index_vector.hpp

```diff
@@ -0,0 +1,148 @@
+    class GroupIterator : public boost::iterator_facade<GroupIterator, Idx, boost::random_access_traversal_tag,
+                                                        boost::iterator_range<IdxCount>, Idx> {
+      public:
+        explicit constexpr GroupIterator(IdxVector const& dense_vector, Idx const& group)
+            : dense_vector_{dense_vector},
+              group_{group},
+              group_range_{std::equal_range(std::cbegin(dense_vector_), std::cend(dense_vector_), group)} {}
+
+      private:
+        using group_iterator = IdxVector::const_iterator;
+
+        IdxVector const& dense_vector_;
+        Idx group_;
+        std::pair<group_iterator, group_iterator> group_range_;
+
+        friend class boost::iterator_core_access;
+
+        boost::iterator_range<IdxCount> dereference() const {
+            return boost::counting_range(static_cast<Idx>(std::distance(std::cbegin(dense_vector_), group_range_.first)),
+                                         static_cast<Idx>(std::distance(std::cbegin(dense_vector_), group_range_.second)));
+        }
+        constexpr bool equal(GroupIterator const& other) const { return group_ == other.group_; }
+        constexpr void increment() { advance(1); }
+        constexpr void decrement() { advance(-1); }
+        constexpr void advance(Idx n) {
+            auto const start = n > 0 ? group_range_.second : std::cbegin(dense_vector_);
+            auto const stop = n > 0 ? std::cend(dense_vector_) : group_range_.first;
+
+            group_ += n;
+            group_range_ = std::equal_range(start, stop, group_);
+        }
+    };
+
+    constexpr auto group_iterator(Idx group) const { return GroupIterator{dense_vector_, group}; }
+
+  public:
+    DenseIdxVector() = default;
+    explicit DenseIdxVector(IdxVector dense_vector, Idx groups_size)
+        : dense_vector_(std::move(dense_vector)), groups_size_(groups_size) {}
+
+    constexpr auto size() const { return groups_size_; }
+    constexpr auto begin() const { return group_iterator(Idx{}); }
+    constexpr auto end() const { return group_iterator(size()); }
+
+    constexpr auto element_size() const { return dense_vector_.size(); }
+    constexpr auto get_group(Idx element) const { return dense_vector_[element]; }
+    auto get_element_range(Idx group) const { return *group_iterator(group); }
+
+  private:
+    IdxVector dense_vector_;
+    Idx groups_size_;
+};
+
+template <typename T>
+concept grouped_idx_vector_type =
+    std::is_same<T, const SparseIdxVector>::value || std::is_same<T, const DenseIdxVector>::value;
+
+template <grouped_idx_vector_type First, grouped_idx_vector_type... Rest>
+auto zip_sequence(First& first, Rest&... rest) {
+    auto all_equal_sizes = ((first.size() == rest.size()) && ...);
+    assert(all_equal_sizes);
+    auto zip_begin = boost::make_zip_iterator(boost::make_tuple(first.begin(), rest.begin()...));
+    auto zip_end = boost::make_zip_iterator(boost::make_tuple(first.end(), rest.end()...));
```

Comment:
```suggestion
    assert((first.size() == rest.size()) && ...);
    auto const zip_begin = boost::make_zip_iterator(boost::make_tuple(first.begin(), rest.begin()...));
    auto const zip_end = boost::make_zip_iterator(boost::make_tuple(first.end(), rest.end()...));
```

Reply:
This was giving a compile error in gcc. It said the fold doesnt work directly within braces of assert.

Reply:
then with an additional pair of braces

```suggestion
    assert(((first.size() == rest.size()) && ...));
    auto const zip_begin = boost::make_zip_iterator(boost::make_tuple(first.begin(), rest.begin()...));
    auto const zip_end = boost::make_zip_iterator(boost::make_tuple(first.end(), rest.end()...));
```